modules/transactions/src/integrationTest/java/org/apache/ignite/internal/table/ItTransactionRecoveryTest.java

```diff
@@ -697,6 +706,119 @@ public void testFinishAlreadyFinishedTx() throws Exception {
         assertThat(finish2, willThrow(TransactionAlreadyFinishedException.class));
     }
 
+    @Test
+    public void testPrimaryFailureRightAfterCommitMsg() throws Exception {
+        TableImpl tbl = (TableImpl) node(0).tables().table(TABLE_NAME);
+
+        var tblReplicationGrp = new TablePartitionId(tbl.tableId(), 0);
+
+        String leaseholder = waitAndGetLeaseholder(node(0), tblReplicationGrp);
+
+        IgniteImpl commitPartNode = commitPartitionPrimaryNode(leaseholder);
+
+        log.info("Transaction commit partition is determined [node={}].", commitPartNode.name());
+
+        IgniteImpl txCrdNode = nonPrimaryNode(leaseholder);
+
+        log.info("Transaction coordinator node is determined [node={}].", txCrdNode.name());
+
+        Transaction rwTx1 = createRwTransaction(txCrdNode);
+
+        CompletableFuture<?> commitMsgSentFut = new CompletableFuture<>();
+        CompletableFuture<?> cancelLeaseFuture = new CompletableFuture<>();
+
+        txCrdNode.dropMessages((nodeName, msg) -> {
+            if (msg instanceof TxFinishReplicaRequest) {
+                boolean isFirst = !commitMsgSentFut.isDone();
+
+                if (isFirst) {
+                    commitMsgSentFut.complete(null);
+
+                    return true;
+                } else {
+                    cancelLeaseFuture.join();
```

Comment:
I have run it many times and it seems not to be flaky. I am not sure that this is an issue. If it appears to be flaky, we can fix it, otherwise we can use this approach